Step1 - 인수 테스트 기반 리팩터링

ssamzag opened the pull request:
안녕하세요. 오태경 리뷰어님!
교육생 최민호라고 합니다.

1단계 리뷰 요청드립니다. 
시나리오 인수 테스트 리팩토링은 진행하다보니 기존 것들과 중복이 되는 것 같아 모두 지우게 되었네요;

앞으로 잘 부탁드립니다. 😄

ssamzag committed:
refactor: LineService 비즈니스로직 -> 도메인 리팩토링

ssamzag committed:
refactor: station() 메서드 while문 depth 줄임

ssamzag committed:
refactor&test: 구간 단건 테스트들을 흐름을 검증하는 테스트로 리팩터링

ssamzag committed:
refactor&test: 지하철 노선 인수테스트 시나리오 기반으로 리팩터링

ohtaeg commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -21,8 +21,8 @@
 public class LineService {
-    private LineRepository lineRepository;
-    private StationService stationService;
+    private final LineRepository lineRepository;
+    private final StationService stationService;
```
final 👍

ohtaeg commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -32,22 +32,15 @@ public LineService(LineRepository lineRepository, StationService stationService)
-                .map(it -> StationResponse.of(it))
-                .collect(Collectors.toList());
-        return LineResponse.of(persistLine, stations);
+        Line persistLine = lineRepository.save(
```
사소하지만 개인적으로 메서드 내에서 영속상태인 Line과 비영속/준영속 상태인 Line과 구분해서 사용해야하는 경우라면
persistLine과 같이 상태를 포함한 네이밍이 좋을 것 같아요 😄
해당 메서드 내에서는 영속상태인 persistLine만 다루고 있기에 단순하게 line으로 명명짓는다면 직관적일 것 같은데 어떻게 생각하시나요 :) ?

ssamzag replied:
말씀하신대로 메서드마다 일관적으로 명명하면 구분하기도 좋아 보이네요. 👍 
적용해 보도록 하겠습니다!!

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -16,8 +15,8 @@ public class Line extends BaseEntity {
     private String name;
     private String color;
```
사소하지만 Line을 만들때 호출되는 생성자들은 다 color를 포함하고 있는데요 :)
그로인해 color 값은 필수 값으로 유추가 되기에 color에다가 nullable = false 속성을 부여하는 것도 좋을 것 같아요 😄

ssamzag replied:
넵! not null 속성 붙일게요!

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -51,6 +50,19 @@ public String getColor() {
     }
 
     public List<Section> getSections() {
-        return sections;
+        return sections.getSections();
+    }
```
요건 사용하지 않는 것 같아요 😄

ssamzag replied:
기계적으로 리팩토링하다보니 넣게 됐네요.. 사용 안 하는 것이니 지우도록 하겠습니다 😃

ohtaeg commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -58,10 +51,7 @@ public Line findLineById(Long id) {
 
     public LineResponse findLineResponseById(Long id) {
```
조회시 `@Transactional(readonly = true)` 를 붙여주셔도 좋을 것 같아요 :)

ssamzag replied:
넵넵 붙이도록 할게요

ohtaeg commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -21,8 +21,8 @@
 @Service
 @Transactional
 public class LineService {
```
비지니스 로직을 잘 옮겨주셨네요 💯

ohtaeg commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -58,10 +51,7 @@ public Line findLineById(Long id) {
+        return LineResponse.of(persistLine);
     }
 
     public void updateLine(Long id, LineRequest lineUpdateRequest) {
```
사소하지만 RuntimeException도 좋지만 어떤 Line을 찾다가 예외가 발생했는지 자세히 알려주는 것은 어떨까 싶어요 ㅎㅎ
운영 단계에서 로그를 찾아볼 때 큰 도움이 되지 않을까 해요!

ssamzag replied:
리팩토링 후다닥 끝내고 다음 미션 진행하고 싶은 마음에.. 🥲  
리팩토리 진행한 모든 예외 다시 찾아서 수정하도록 할게요!

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,159 @@
+            throw new RuntimeException("이미 등록된 구간 입니다.");
+        }
+
+        if (!stations().isEmpty() && !isPresentStation(upStation) && !isPresentStation(downStation) ) {
```
다양한 식으로 이루어진 조건문이라면 메서드로 분리하여 가독성을 높이는 방법도 괜찮을 것 같은데
민호님은 어떻게 생각하시나요 😄

ssamzag replied:
확실히 한 눈에 안 읽히네요. 메서드로 분리하도록 할게요!!

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    public List<Section> getSections() {
+        return Collections.unmodifiableList(sections);
```
컬렉션을 외부에서 변조할 수 없도록 👍

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    private Station firstStation() {
+        Station downStation = sections.get(0).getUpStation();
```
의미만 놓고 보았을때, 첫번째 구간에서 상행역을 가져오는 코드로 이해가 되지만 네이밍은 하행역을 가져오도록 되어있는데요!
아래 로직도 같이봐야 downStation이라는 네이밍이 이해가 가지만 다른 개발자분이 첫 코드만 봤을 때 오해할 수도 있는 부분 같아요 :)

ssamzag replied:
이것도 기계적 리팩토링의 부작용이네요 😢 
변수이름까지 꼼꼼하게 체크하도록 할게요!

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,159 @@
+
+    private boolean isPresentPreSection(Station station) {
+        return sections.stream()
+                .filter(s -> s.getUpStation() != null)
```
상행역을 직접 가져와도 좋지만, Section에게 직접 상행역이 비어있는지 아닌지 물어보면 어떨까요 :) ?

ssamzag replied:
넵. 수정하도록 할게요!!

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,159 @@
+
+    private boolean isPresentNextSection(Station station) {
+        return sections.stream()
+                .filter(s -> s.getUpStation() != null)
```
같은 피드백 입니다 😄 
https://github.com/next-step/atdd-subway-service/pull/591#discussion_r896694750

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,159 @@
+        }
+
+        if (!stations().isEmpty() && !isPresentStation(upStation) && !isPresentStation(downStation) ) {
+            throw new RuntimeException("등록할 수 없는 구간 입니다.");
```
같은 피드백 입니다 😄 
https://github.com/next-step/atdd-subway-service/pull/591#discussion_r896680078

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,159 @@
+                .findFirst();
+    }
+
+    private void createSection(Line line, Optional<Section> upLineStation, Optional<Section> downLineStation) {
```
소나큐브에서는 Optional의 탄생 배경은 Optional은 반환 타입으로만 사용되도록 의도한 설계라고 하며,
파라미터로 Optional 사용을 지양하라고 하는데요 :)
https://rules.sonarsource.com/java/RSPEC-3553

개인적으로도 메서드의 파라미터가 인스턴스일 때 Optional이든 아니든
null 체크를 메서드 내부에서 항상 하는 것이 언제나 안전하기에
인스턴스를 Optional로 한번 더 감쌀 이유가 있을지 개인적으로 항상 고민을 하는 부분이에요 :)

민호님은 어떻게 생각하시나요? 😄

ssamzag replied:
Optional에 그런 설계가 의도되어 있었군요. 좋은 정보를 알게되었습니다. 👍 
적어주신 링크와 말씀해주신 것을 읽으니 넘길 이유가 없겠네요.
수정하도록 하겠습니다!

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,159 @@
+        if (!upLineStation.isPresent() || !downLineStation.isPresent()) {
+            return;
+        }
+        int newDistance = upLineStation.get().getDistance() + downLineStation.get().getDistance();
```
[객체지향 생활 체조 원칙](https://developerfarm.wordpress.com/2012/01/27/object_calisthenics_4/)에서 언급하고 있지만 int 값 자체는 아무 의미 없는 스칼라 값일 뿐이라고 표현을 하고 있습니다..!

아무래도 해당 표현에서 전달하고자 하는 의미는
원시형 데이터 자체는 큰 의미를 담고있지 않기에 그저 변수명으로만 의미를 추론할 수 있는데요
원시값을 객체화 시킨다면 개발자에게 해당 값이 어떤 값이며 왜 쓰는지에 대해 정보를 보다 명확하게 전달할 수 있다고 표현하는 것 같아요..!

게다가 Sections 내부에서 distance를 얻어와서 거리를 증가시키는 로직도 위임하는 장점도 있을 것 같아보여서 
동의가 되는 부분이기에 민호님의 의견은 어떠하신지 궁금하네요 😄 

물론 해당 피드백은 개인적인 의견이라 꼭 수정해주실 필요가 없습니다 :)

ssamzag replied:
RacingCar 미션하면서 원시값 포장이라는 것을 처음 알게 되었는데 만들 때는 귀찮지만 한 번 구현해 놓으면 그 다음부터는 정말 편하긴 하더라구요. 
이번 것도 깔끔한 코드를 위해 적용하도록 해야겠네요. 감사합니다!

ohtaeg commented on src/test/java/nextstep/subway/line/acceptance/LineAcceptanceTest.java:
```diff
@@ -39,82 +40,70 @@ public void setUp() {
     }
 
-    @DisplayName("지하철 노선을 생성한다.")
+    /**
+     * Given 지하철 역이 등록 되어 있고
+     * When 지하철 노선을 등록하면
+     * Then 지하철 노선이 등록된다
+     *
+     * When 지하철 노선을 조회하면
+     * Then 지하철 노선이 조회된다
+     *
+     * When 지하철 노선 목록을 조회요청하면
+     * Then 지하철 노선 목록이 조회된다
+     *
+     * When 기존에 존재하는 지하철 노선 이름으로 지하철 노선을 생성하면
+     * Then 생성에 실패한다
+     *
+     * When 지하철 노선을 수정하면
+     * Then 지하철 노선이 수정된다
+     *
+     * When 지하철 노선을 삭제하면
+     * Then 지하철 노선이 삭제된다
+     *
+     * When 지하철 노선을 조회하면
+     * Then 등록된 노선은 0개이다
+     */
+    @DisplayName("지하철 노선 관련 시나리오 테스트")
```
시나리오 테스트 👍

ohtaeg commented on src/test/java/nextstep/subway/line/domain/SectionsTest.java:
```diff
@@ -0,0 +1,95 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class SectionsTest {
```
Sections 테스트를 잘해주셨네요 👍 
역 사이에 새로운 역을 등록하는 케이스도 추가 작성해주시면 좋을 것 같아요 :)

ssamzag replied:
넵 추가할게요!

ohtaeg commented on src/test/java/nextstep/subway/line/domain/SectionsTest.java:
```diff
@@ -0,0 +1,95 @@
+    @BeforeEach
+    void setUp() {
+        station1 = new Station("강남역");
+        station2 = new Station("력삼역");
```
력 💪

ssamzag replied:
💪력!

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -4,8 +4,7 @@
+import java.util.*;
 
 @Entity
 public class Line extends BaseEntity {
```
번거로우시겠지만 Line 및 Section에 대한 도메인 테스트도 추가해주시면 감사하겠습니다 🙇

ssamzag replied:
넵 추가하도록 할게요!

ohtaeg reviewed (changes requested):
안녕하세요 민호님!
이번 ATDD - TDD 미션을 같이 함께할 리뷰어 오태경이라고 합니다 👋 
함께하게 되어 영광이며, 저도 잘 부탁드립니다 🙇 

구조적인 리팩토링을 잘해주셨어요 👍 👍 
미션 요구사항을 잘 만족시켜주셨는데, 리뷰를 드리다보니 사소하게 많이 드린 것 같아요.!
다음 미션을 진행하시기전에 확인해주시고 반영해주실 부분은 반영해주시면 좋을 것 같아
Request changes 하였습니다 🙏 

확인해주시고 재요청 주시면 감사하겠습니다 :)

ssamzag committed:
chore: 클래스 파일 추가

ssamzag committed:
feat&test: color 필드 not null 속성 추가 및 Line 도메인 테스트 추가

ssamzag committed:
style: 클래스 DisplayName 변경

ssamzag committed:
refactor: distance 객체포장 외
- 예외 상세 메시지 추가
- 영속상태가 단일로 있는 메서드의 변수명 persist 프리픽스 제거
- findLineResponseById transaction readonly 속성 추가
등

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -0,0 +1,45 @@
+import javax.persistence.Embeddable;
+import java.util.Objects;
+
+@Embeddable
+public class Distance {
```
피드백 반영 💯

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -17,30 +17,42 @@ public Sections() {
+        return isPresentStation(upStation) && isPresentStation(downStation);
+    }
+
+    private boolean canAdd(Station upStation, Station downStation) {
+        return !stations().isEmpty() && !isPresentStation(upStation) && !isPresentStation(downStation);
```
👍 👍

ohtaeg commented on src/test/java/nextstep/subway/line/domain/DistanceTest.java:
```diff
@@ -0,0 +1,31 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.*;
+
+class DistanceTest {
```
꼼꼼한 도메인 테스트 👍

ohtaeg commented on src/test/java/nextstep/subway/line/domain/SectionsTest.java:
```diff
@@ -28,12 +31,28 @@ void setUp() {
     }
 
+    @Test
+    @DisplayName("강남역 상행으로 서초역을 등록한다")
+    void addUpStation() {
+        sections.add(line, station0, station1, 40);
+        assertThat(sections.stations()).hasSize(3);
+        assertThat(sections.stations()).contains(station0);
+    }
+
+    @Test
+    @DisplayName("선릉역 하행으로 삼성역을 등록한다")
```
구간 추가 테스트 👍

ohtaeg reviewed (approved):
피드백 반영을 잘해주셨네요 💯 
사소한 코멘트가 많았는데 다 적용해주셔서 감사드립니다 🙇 
비가 와서 저녁에 추울 수도 있으니 몸 조심하시구요 :)
다음 미션도 화이팅입니다! 력! 💪
